### PR TITLE
feat(medusa): Allow upsert for the optional cart metadata field

### PR DESCRIPTION
**Pull request by mirko-pira**
### What
This PR allows using the optional metadata field for the create-cart and update-cart requests in the store/carts module.

### Why
The metadata field can be used to include additional information with the request, such as tracking information or customer-specific details. This makes the create-cart and update-cart requests more flexible and customizable, and can improve the overall user experience.

### How
The metadata field has been added to the request body for both create-cart and update-cart requests. It has also been included in the API documentation and validation schema.

### Testing
This change has been tested locally by creating and updating carts with and without the metadata field. Additionally, unit tests have been added to ensure that the metadata field is properly validated and processed. Reviewers can also test the feature by creating and updating carts with and without the metadata field and verifying that the expected behaviour occurs.

### CODE REVIEW

File: packages/medusa/src/api/routes/store/carts/create-cart.ts [hunk 1]
```diff
@@ -241,6 +241,9 @@ export class StorePostCartReq {
   @IsOptional()
   context?: object
 
+  @IsOptional()
```
Comment (olivermrbl) [inline packages/medusa/src/api/routes/store/carts/create-cart.ts:244]: todo: Add property to `@schema StorePostCartReq`

File: packages/medusa/src/api/routes/store/carts/update-cart.ts [hunk 2]
```diff
@@ -219,6 +219,9 @@ export class StorePostCartsCartReq {
   @IsOptional()
   context?: object
 
+  @IsOptional()
+  metadata?: Record<string, unknown>
```
Comment (olivermrbl) [inline packages/medusa/src/api/routes/store/carts/update-cart.ts:223]: todo: Add property to `@schema StorePostCartsCartReq`

**Review by olivermrbl — CHANGES_REQUESTED** [review]
Great little improvement - thanks for the contribution. And props for the commit messages. Prettiest I've seen in a while 😍

File: packages/medusa/src/api/routes/store/carts/create-cart.ts (hunk 1 above)
Comment (mirko-pira) [reply packages/medusa/src/api/routes/store/carts/create-cart.ts:249]: Done

File: packages/medusa/src/api/routes/store/carts/update-cart.ts (hunk 2 above)
Comment (mirko-pira) [reply packages/medusa/src/api/routes/store/carts/update-cart.ts:228]: Done ✅

File: packages/medusa/src/services/cart.ts [hunk 3]
```diff
@@ -313,6 +313,7 @@ class CartService extends TransactionBaseService {
 
         const rawCart: DeepPartial<Cart> = {
           context: data.context ?? {},
+          metadata: data.metadata ?? {}
```
Comment (adrien2p) [inline packages/medusa/src/services/cart.ts:316]: todo: this can be removed, it is manage l418-431

Comment (adrien2p) [reply packages/medusa/src/services/cart.ts:316]: suggestion: rm line 421 and after before the remaining fields do a check if data.metadata is present and use the setMetadata utils like in the other places

**Review by adrien2p — CHANGES_REQUESTED** [review]

File: packages/medusa/src/services/cart.ts (hunk 3 above)
Comment (mirko-pira) [reply packages/medusa/src/services/cart.ts]: Done ✅
Thanks for the suggestion.

**Review by adrien2p — APPROVED** [review]
LGTM

**Comment by olivermrbl** [comment]
@mirko-pira Please correct tests, so we can merge :)

**Comment by mirko-pira** [comment]
@olivermrbl I made the change and the unit tests passed. However, for some reason, the integration tests are not working in my local environment.

image

**Comment by riqwan** [comment]
Hey, thanks for the PR! Since v2 brought a lot of architectural and API changes on the backend, we will be closing this ticket.

**Pull request closed without merging**